psi4/src/psi4/libscf_solver/sad.cc

```diff
@@ -247,14 +252,85 @@ SharedMatrix SADGuess::form_D_AO() {
+        }
+
+        // Occupation numbers
+        SharedVector occ_a, occ_b;
+        // Number of orbitals occupied, partially or fully
+        int nocc_a, nocc_b;
+        if (options_.get_bool("SAD_FRAC_OCC")) {
+            // The density is spread over the whole of the possible valence shell.
+            // Only the noble gas core is doubly occupied
+            static const std::vector<int> magic_values = {0, 2, 10, 18, 36, 54, 86, 118};
+            // Find the noble gas core.
+            auto imagic = std::lower_bound(magic_values.begin(), magic_values.end(), Z);
+            if (imagic == magic_values.end()) {
+                throw PSIEXCEPTION("SAD: Fractional occupations are not supported beyond Oganesson");
+            }
+            // lower_bound gives a value that is equal or greater than the
+            // wanted value, which is handled next.
+            if (*imagic > Z) imagic--;
+
+            // Number of frozen and active orbitals
+            int nfzc, nact;
+            if ((*imagic) == Z) {
+                // Special case: we can hit the boundary at the end of the array
+                nfzc = 0;
+                nact = (*imagic) / 2;
+            } else {
+                nfzc = (*imagic) / 2;
+                nact = (*(++imagic)) / 2 - nfzc;
+            }
+
+            // Number of occupied orbitals is
+            nocc_a = nocc_b = nfzc + nact;
+
+            // Fractional alpha and beta occupation. Occupations are
+            // squared in the density calculation, so take the root
+            double frac_a = std::sqrt((nalpha[index] - nfzc) / nact);
+            double frac_b = std::sqrt((nbeta[index] - nfzc) / nact);
+
+            occ_a = std::make_shared<Vector>("Alpha fractional occupation", nocc_a);
+            for (size_t x = 0; x < nfzc; x++) occ_a->set(x, 1.0);
+            for (size_t x = nfzc; x < nocc_a; x++) occ_a->set(x, frac_a);
+
+            occ_b = std::make_shared<Vector>("Beta fractional occupation", nocc_b);
+            for (size_t x = 0; x < nfzc; x++) occ_b->set(x, 1.0);
+            for (size_t x = nfzc; x < nocc_b; x++) occ_b->set(x, frac_b);
+
+            if (print_ > 1) {
+                outfile->Printf(
+                    "  %d fully and %d partially occupied orbitals with % .3f alpha and % .3f beta electrons "
+                    "each\n",
+                    nfzc, nact, (nalpha[index] - nfzc) / nact, (nbeta[index] - nfzc) / nact);
+            }
+        } else {
+            // Conventional occupations
+            nocc_a = std::round(nalpha[index]);
+            occ_a = std::make_shared<Vector>("Alpha occupation", nocc_a);
+            for (size_t x = 0; x < nocc_a; x++) occ_a->set(x, 1.0);
+
+            nocc_b = std::round(nbeta[index]);
+            occ_b = std::make_shared<Vector>("Beta occupation", nocc_b);
+            for (size_t x = 0; x < nocc_b; x++) occ_b->set(x, 1.0);
```

Comment:
Is the history incorrect here, this should be merged with another PR no?

Reply:
It **is** correct - if you look down in the file, you'll see that this code has just been refactored. It used to be inside the atomic solver, but it's simpler to determine the occupations along with the spin state in the driver and just pass the occupation arrays to the solver since it's all it needs anyway.